nemo/collections/nlp/modules/common/megatron/adapters/mcore_mixins.py

```diff
@@ -0,0 +1,66 @@
+import torch
+from megatron.core.models.gpt.gpt_embedding import GPTEmbedding
+from megatron.core.transformer.custom_layers.transformer_engine import TELinear
+
+from nemo.collections.nlp.modules.common.megatron.adapters.parallel_adapters import (
+    AdapterName,
+    LoraKQVAdapterConfig,
+    PromptEncoderAdapterConfig,
+)
+from nemo.core import adapter_mixins
+from nemo.utils import model_utils
+
+
+def swap_mcore_mixin(module, mcore_mixin):
+    """
+    Casts module to mcore_mixin and register corresponding adapters.
+    """
+    module.__class__ = mcore_mixin
+    module.mcore_register_adapters()
+
+
+class McoreAdapterModuleMixin(adapter_mixins.AdapterModuleMixin):
+    def mcore_register_adapters(self):
+        raise NotImplementedError("Mcore mixins should implement setup_adapters on a subclass of MyBase")
+
+
+class LoraTELinear(TELinear, McoreAdapterModuleMixin):
```

Comment:
Could we use the [SelfAttention class from Megatron-LM ](https://github.com/NVIDIA/Megatron-LM/blob/f24fac4ed0dcf0522056521a93445d9a82f501a9/megatron/core/transformer/attention.py#L245) and implement the [get_query_key_value_tensors](https://github.com/NVIDIA/Megatron-LM/blob/f24fac4ed0dcf0522056521a93445d9a82f501a9/megatron/core/transformer/attention.py#L266C20-L266C20) here instead of going to the lower level TELinear? It is a few more lines of code but is gives me a more general recipe for implementing future PEFT methods. LoRA is a bit of an outlier because it can be fully encapsulated into a low-level Linear layer.

Reply:
I've seen people injecting lora on not only `linear_qkv` but also other linears in the attention, although nemo only applies lora on `linear_qkv` for now...